File: sach/de3366.py
//DE3366.PY - Tran Huu Nam - hsgtin.vn - 2024
import sys
n, k = map(int, input().split())
s = input().strip()
s += s
f = [[[[False] * (2 * n + 1) for _ in range(101)] for _ in range(k)] for _ in range(k)]
for d in range(n):
    for i in range(2 * n - d):
        j = i + d
        for x in range(k):
            for y in range(k):
                x2 = x + (s[i] == 'B')
                if x2 < k and not f[y][x2][i + 1][j]:
                    f[x][y][i][j] = True
                x2 = x + (s[j] == 'B')
                if x2 < k and not f[y][x2][i][j - 1]:
                    f[x][y][i][j] = True
dem = 0
for i in range(n):
    if f[0][0][i][i + n - 1]:
        dem += 1
print(dem)

File: sach/de3366.cpp
//DE3366.CPP - Tran Huu Nam - hsgtin.vn - 2024
#include <bits/stdc++.h>
using namespace std;
int n,k;
string s;
bool f[31][31][101][201]={};
int main() {
    cin>>n>>k>>s;
    s+=s;
    int x,y,x2,i,j,d;
    for (d=0; d<n; d++) {
      for (i=0; i<2*n-d; i++) {
        j=i+d;
        for (x=0; x<k; x++) {
          for (y=0; y<k; y++) {
            //f[x][y][i][j]=false;
            x2=x + (s[i]=='B');
            if (x2<k && f[y][x2][i+1][j]==false)
                f[x][y][i][j]=true;
            x2=x + (s[j]=='B');
            if (x2<k && f[y][x2][i][j-1]==false)
                f[x][y][i][j]=true;
          }
        }
      }
    }
    int dem=0;
    for (i=0; i<n; i++)
      if (f[0][0][i][i+n-1]) dem++;
    cout<<dem;
}

